billing/lib/api.ts

```diff
@@ -0,0 +1,290 @@
+
+/**
+ * Captures details about a customer that should be billed for a given period
+ * of usage.
+ */
+export interface CustomerBillingInstruction {
+  /** Customer DID (did:mailto:...). */
+  customer: DID<'mailto'>
+  /**
+   * Opaque identifier representing an account in the payment system.
+   * 
+   * e.g. Stripe customer ID (stripe:cus_9s6XKzkNRiz8i3)
+   */
+  account: string
+  /** Unique identifier of the product a.k.a tier. */
+  product: string
+  /** Time the billing period spans from (inlusive). */
+  from: Date
+  /** Time the billing period spans to (exclusive). */
+  to: Date
+}
+
+export type CustomerBillingQueue = QueueAdder<CustomerBillingInstruction>
+
+/**
+ * Captures details about a space that should be billed for a given customer in
+ * the given period of usage.
+ */
+export interface SpaceBillingInstruction extends CustomerBillingInstruction {
+  /** Space DID (did:key:...). */
+  space: DID
+  /** Storage provider for the space. */
+  provider: DID<'web'>
+}
+
+export type SpaceBillingQueue = QueueAdder<SpaceBillingInstruction>
+
+// Upload API stores //////////////////////////////////////////////////////////
+
+export interface Consumer {
+  consumer: DID
+  provider: DID<'web'>
+  subscription: string
+  cause: Link
+  insertedAt: Date
+  updatedAt: Date
+}
+
+export interface ConsumerKey { subscription: string, provider: DID<'web'> }
+export interface ConsumerListKey { consumer: DID }
+
+export type ConsumerStore =
+  & StoreGetter<ConsumerKey, Consumer>
+  & StoreLister<ConsumerListKey, Pick<Consumer, 'consumer'|'provider'|'subscription'>>
+
+export interface Subscription {
+  customer: DID<'mailto'>
+  provider: DID<'web'>
+  subscription: string
+  cause: Link
+  insertedAt: Date
+  updatedAt: Date
+}
+
+export interface SubscriptionKey { provider: DID<'web'>, subscription: string }
+export interface SubscriptionListKey { customer: DID<'mailto'> }
+
+export type SubscriptionStore =
+  & StoreGetter<SubscriptionKey, Subscription>
+  & StoreLister<SubscriptionListKey, Pick<Subscription, 'customer'|'provider'|'subscription'|'cause'>>
+
+// UCAN invocation ////////////////////////////////////////////////////////////
+
+// TODO: replace `UcanInvocation` type in `ucan-invocation/types.ts` with this?
+export interface UcanMessage<C extends Capabilities = Capabilities> {
+  carCid: Link
+  invocationCid: Link
+  value: UcanMessageValue<C>
+  ts: Date
+}
+
+export interface UcanMessageValue<C extends Capabilities = Capabilities> {
+  att: C,
+  aud: DID,
+  iss?: DID,
+  prf?: Array<LinkJSON<Link>>
+}
+
+export interface UcanReceiptMessage<
+  C extends Capabilities = Capabilities,
+  R extends Result = Result
+> extends UcanMessage<C> {
+  type: 'receipt'
+  out: R
+}
+
+export interface UcanWorkflowMessage<C extends Capabilities = Capabilities> extends UcanMessage<C> {
+  type: 'workflow'
+}
+
+export type UcanStreamMessage<C extends Capabilities = Capabilities> = UcanWorkflowMessage<C> | UcanReceiptMessage<C>
```

Comment:
Something about these types are kind of off, [here is definition in ucanto](https://github.com/web3-storage/ucanto/blob/43ea497ca2200f9e1e15dbbdd033ec24895edb7d/packages/interface/src/lib.ts#L802-L821) and few things to note:

1. We could have more than one invocation / receipt in a single message.
2. Message can contain both invocation and receipt even though often it's either on or another.
3. We intentionally did not use term `workflow` because IPVM calls it something else.

Probably none of the above is critical, but might be worth considering.

Reply:
Yes agreed they are not great. They are an improvement on https://github.com/web3-storage/w3infra/blob/fe3e3c6f3385e2b519eb1b296c2f3de0e12d1390/ucan-invocation/types.ts#L47-L76 which would have needed a whole lot of type casting / use of `any` to be usable. Unfortunately the UCAN stream is coupled to the current invocations as they exist in ucanto plus our hacked in receipts. I really don't want to increase the scope of this PR to fix this, but propose instead that when we implement the invocation spec we do it.